### PR TITLE
ARROW-14182: [C++][Compute] Hash Join performance improvement

### PR DESCRIPTION
**Pull request by michalursa**
Faster implementation of hash join.

### CODE REVIEW

**Comment by nealrichardson** [comment]
@save-buffer PTAL

**Comment by westonpace** [comment]
Can we rebase this now that the bloom filter (and key_hash) changes have been merged in.

File: cpp/src/arrow/compute/exec/swiss_join.h [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+                        const uint16_t* row_ids, int* num_appended, int num_cols,
+                        const int* col_ids = NULLPTR);
+
+  Status AppendNulls(MemoryPool* pool,
+                     const std::vector<std::shared_ptr<DataType>>& types,
+                     int num_rows_to_append);
+
+  Status AppendNulls(MemoryPool* pool,
+                     const std::vector<std::shared_ptr<DataType>>& types,
+                     int num_rows_to_append, int* num_appended);
```
Comment (zagto) [inline cpp/src/arrow/compute/exec/swiss_join.h:94]: It looks like this overload is no longer used anywhere

**Comment by westonpace** [comment]
@michalursa I think this can be closed in favor of #13493

**Pull request closed without merging**